pkg/events/emitter.go

```diff
@@ -0,0 +1,54 @@
+package events
+
+import (
+	"sync"
+)
+
+type Event[T any] interface {
+	AddListener(listener Listener[T]) Disposable
+	Fire(data T)
+}
+
+type Disposable func()
+
+type Listener[T any] func(event T)
+
+type listenerWrapper[T any] struct {
+	Listener Listener[T]
+}
+
+type event[T any] struct {
+	listeners  map[int]*listenerWrapper[T]
+	eventMutex sync.Mutex
+	keyCounter int
+}
+
+func NewEvent[T any]() Event[T] {
+	return &event[T]{
+		listeners: make(map[int]*listenerWrapper[T]),
+	}
+}
+
+func (e *event[T]) AddListener(listener Listener[T]) Disposable {
+	e.eventMutex.Lock()
+	e.keyCounter++
+	key := e.keyCounter
+	e.listeners[key] = &listenerWrapper[T]{
+		Listener: listener,
+	}
+	e.eventMutex.Unlock()
+	return func() {
+		e.eventMutex.Lock()
+		defer e.eventMutex.Unlock()
+
+		delete(e.listeners, key)
+	}
+}
+
+func (e *event[T]) Fire(event T) {
+	e.eventMutex.Lock()
+	defer e.eventMutex.Unlock()
```

Comment:
is the locking here needed, we are only accessing the map and not adding/removing elements from it